### ISSUE TITLE
【Task.34】Deterministic Trajectory Replay - RFC

### ISSUE THREAD

**Issue by shanyulu** [redai-studio/Relax#175]
## Proposal

| Item | Value |
|---|---|
| Task | No.034 — Deterministic Trajectory Replay |
| Proposer | @shanyulu |
| Baseline | `redai-infra/Relax@8a54679e971087566bfda939a5f75649e07fb861` |
| Branch | `task/no034-deterministic-trajectory-replay` |

## Summary

Add a versioned replay bundle and a CPU runner for reward, advantage, and loss debugging. A captured step can be replayed without Ray, SGLang, rollout workers, model weights, or GPUs.

```text
R0  verify bundle       manifest + hashes + contracts     → trusted input
R1  replay reward       reward inputs                     → raw/processed reward
R2  replay advantage    reward/KL/value + full cohort     → advantages/returns
R3  replay loss         token statistics + masks          → loss/metrics
──────────────────────────────────────────────────────────────────────────
R4  execute model       checkpoint + forward/backward     → logits/gradients
```

V1 implements R0–R3. R4 is out of scope.

A trajectory alone is insufficient for loss replay: current-policy log-probabilities, entropy, and values come from model forward. Saving full logits is too large. V1 therefore captures the compact post-forward vectors consumed by the loss.

## V1 scope

| Area | V1 | Later |
|---|---|---|
| Algorithms | GRPO, PPO | SAPO, CISPO, GSPO, OPSM |
| Reward | built-in/local deterministic | remote, nondeterministic, arbitrary custom |
| Inputs | text | multimodal, Agentic-specific |
| Parallelism | CP=1; complete DP cohort | CP>1; recorded-statistics DP |
| Selection | sample, batch, step | query/range tooling |
| Evidence | synthetic fixture; one GRPO run and one PPO run | wider recipes |
| Schema | strict v1; incompatible versions reject | migrations, cross-version replay |

No Controller/Service redesign, scheduling or checkpoint changes, public API deletion, unrelated recipe changes, or new dependency.

## Why this boundary is feasible

The baseline already separates the required stages:

| Stage | Production path |
|---|---|
| Reward | `relax/utils/utils.py::post_process_rewards` |
| Train data | `relax/utils/utils.py::convert_samples_to_train_data` |
| Advantage | `relax/components/advantages.py`; `relax/backends/megatron/loss.py` |
| RL kernels | `relax/utils/training/ppo_utils.py` |
| Policy/value loss | `relax/backends/megatron/loss.py` |

Current debug dumps do not define a replay contract: they lack schema/version checks, code/config fingerprints, dependency cohorts, checksums, atomic completion, and current-forward log-probability/entropy capture.

A baseline CPU probe confirms that production `compute_policy_loss` and all `compute_approx_kl` variants run directly from token-level tensors without Megatron initialization.

Independent evidence from [RFC #171](https://github.com/redai-infra/Relax/issues/171) used Qwen3-0.6B GRPO (`rollout_batch_size=4`, `n_samples_per_prompt=8`, `global_batch_size=32`) and recomputed policy loss on CPU:

| Rollout | Offline | Online | Abs. error | Rel. error |
|---:|---:|---:|---:|---:|
| 0 | 0.09789314894454393 | 0.09789315047329218 | 1.53e-9 | 1.56e-8 |
| 1 | 0.08448249922591965 | 0.08448249486510825 | 4.36e-9 | 5.16e-8 |

That test covers `ratio ≡ 1`. The same report observed `train_rollout_logprob_abs_diff=1.22e-2` when old log-probabilities came from SGLang. V1 must therefore capture current-forward log-probabilities for the general `ratio ≠ 1` path.

The numbers above are cited as independent evidence. V1 will produce its own GRPO and PPO comparisons.

### Reward

Capture the fields consumed by the selected reward: prompt, response, label, tokens, response length, status, required metadata, `group_index`, `rm_type`, `reward_key`, post-processing config, and expected raw/processed reward.

Selecting one sample from a group-dependent reward loads its complete original group. Unsupported reward implementations fail explicitly.

### Advantage

Capture rewards, masks, lengths, the selected old-policy source, reference log-probabilities/values when required, estimator/KL/PPO config, complete group/DP cohort, and expected KL/advantages/returns.

The component and Megatron paths are separate producer contracts; replay cannot substitute one for the other. V1 requires every DP shard used by normalization.

### Loss

Capture current response-token log-probabilities, entropy, PPO values, old/rollout/reference log-probabilities, advantages, returns, masks, lengths, reduction config, and expected loss terms.

For `T` response tokens, fp32 `log_probs[T] + entropy[T]` costs `8 × T` bytes. Full logits cost `T × vocab_size × dtype_bytes`.

R3 reproduces scalar policy/value/KL/entropy terms, not gradients. CP>1 fails with a recapture/capability error in V1.

## Bundle

```text
trajectory-replay-v1/
├── manifest.json
├── index.json
├── payload/
│   └── step-000120/
│       ├── reward-batch-000003.pt
│       ├── advantage-cohort-000003-rank-000.pt
│       ├── loss-microbatch-000007-rank-000.pt
│       └── expected.pt
└── COMPLETE
```

The manifest records:

- schema `1.0`, bundle ID, producer commit and dirty-patch digest;
- stage contracts and an allowlisted numerical config hash;
- group/DP cohort identity;
- required/optional fields with dtype, shape and semantic axes;
- payload size and SHA-256;
- selection/dependency expansion and tolerances.

The full argument namespace is never saved.

V1 loading rules:

1. unsupported schema or contract: reject;
2. missing/unknown core field: reject with JSON path;
3. missing DP shard: reject with cohort/rank;
4. checksum, dtype, shape, count, or finiteness mismatch: reject before replay;
5. no silent field drop and no migration in V1.

Payloads contain detached CPU tensors and a strict set of JSON-compatible primitives. Loading uses restricted/weights-only mode where supported plus recursive type validation.

The writer uses a temporary sibling directory, writes and hashes payloads, writes the manifest/index and `COMPLETE`, then atomically renames. A directory without `COMPLETE` is invalid.

## Capture and selection

Capture is off by default. When enabled it selects contract-required tensors, stages them in bounded CPU memory, and performs serialization/hashing outside numerical kernels.

No unconditional `.item()`, `.tolist()`, tensor printing, or GPU synchronization is added to the normal path. A failed capture emits a warning/metric and never publishes a partial bundle; strict debug mode may fail the run.

Selection preserves dependencies:

```json
{
  "requested_samples": [7],
  "captured_samples": [4, 5, 6, 7],
  "dependency_reason": "reward_group=1"
}
```

The report may show sample 7 only, but replay still uses the original group and DP cohort.

## Runner and diagnostics

```bash
python -m relax.tools.trajectory_replay verify --bundle <dir>
python -m relax.tools.trajectory_replay run --stage reward --bundle <dir> --step 120
python -m relax.tools.trajectory_replay run --stage advantage --bundle <dir> --step 120 --sample 7
python -m relax.tools.trajectory_replay run --stage loss --bundle <dir> --step 120 --batch 3
python -m relax.tools.trajectory_replay run --stage all --bundle <dir> --step 120
```

`verify` performs no numerical replay. `run` requires `--stage {reward,advantage,loss,all}`; there is no implicit default. Sample/batch selectors filter the report while preserving dependency closure.

| Exit | Meaning |
|---:|---|
| 0 | pass |
| 1 | numerical/exact mismatch |
| 2 | invalid bundle |
| 3 | unsupported contract or incomplete cohort |
| 4 | reward execution failure |

Initial tolerances, to be calibrated by the GRPO and PPO evidence runs:

| dtype | atol | rtol |
|---|---:|---:|
| fp64 | `1e-10` | `1e-8` |
| fp32 | `1e-6` | `1e-5` |
| fp16/bf16 | `5e-3` | `5e-3` |

Example failure:

```text
FAIL stage=loss/policy-v1 step=120 batch=3 rank=0 sample=7
field=loss_masks token=418 expected=1 actual=0
pg_loss expected=0.083194 actual=0.079802 abs=0.003392 rel=0.04077
suggestion=inspect truncation/mask construction before policy reduction
```

Floating reports include first mismatch, maximum absolute/relative error, mismatch count, and non-finite count.

## Verification

| Area | Coverage |
|---|---|
| Bundle | round-trip, unsupported version, missing field/shard, truncation, checksum, dtype/shape/count, NaN/Inf, absent `COMPLETE`, unsafe object |
| Selection | sample + group closure, batch, step, complete DP cohort |
| Reward | scalar/dict + `reward_key`, metadata, group normalization, execution failure |
| Advantage | KL variants, GRPO returns, PPO GAE, masks/truncation, DP normalization |
| Loss | PPO clipping, entropy/KL, value clipping, reduction modes, all-masked, `ratio ≠ 1` |

A deterministic fixture injects four independent failures: one reward, one mask token, one old log-probability, and one required schema field. Tests assert the stage, sample, field/token, and exit code. One case becomes the debugging walkthrough.

One small GRPO run and one small PPO run each report online/offline raw reward, processed reward, KL, advantages, returns, loss terms, reduction numerator/denominator, and absolute/relative errors. Only synthetic fixtures are committed; real bundles remain external test artifacts.

## Performance

Before a prototype, performance numbers are budgets rather than acceptance criteria:

- disabled-path overhead: initial target `<0.5%`;
- selected-step capture overhead: initial target `<3%`.

PR evidence will use three repeats or three stable windows and report step latency, bundle size, write time, and peak host memory. Final thresholds will be set from measured baseline variance.

### PR 1 — Offline core

- strict v1 schema, hashes and atomic I/O;
- sample/batch/step index with dependency closure;
- CPU GRPO/PPO replay;
- injected-error fixture and tests.

No Megatron hot-path or global argument changes.

### PR 2 — Capture and evidence

- narrow reward/advantage/loss capture hooks;
- current log-probability/entropy/value export;
- complete DP cohort capture;
- GRPO/PPO comparisons, performance table, and debugging guide.

Capture remains off by default. Replay modules are additive; capture failure cannot alter rollout, reward, loss, or optimizer outputs.

Expected scope:

- `relax/utils/training/trajectory_replay/`
- `relax/tools/trajectory_replay.py`
- narrow reviewed reward/advantage/loss hooks;
- `tests/utils/training/trajectory_replay/`
- `tests/fixtures/trajectory_replay/`
- schema and debugging docs.

## Acceptance

| Official requirement | V1 evidence |
|---|---|
| Minimal boundary | R0–R3; model forward/backward excluded |
| Key tensors reproduced | GRPO and PPO exact/dtype-aware comparisons |
| Sample, batch, step | indexed selectors with dependency closure |
| Schema failure | strict v1 rejection; no silent drop |
| Failure diagnosis | injected-error fixture and guide |
| No rollout cluster | CPU runner imports neither Ray nor SGLang |

## Environment

| Item | Value |
|---|---|
| OS | Ubuntu 22.04.5 LTS / Linux 5.15, x86_64 |
| Python | 3.12.3 |
| Offline dependencies | PyTorch 2.8.0+cu128, CUDA runtime 12.8; pytest 8.4.2 |
| Capture environment | Relax baseline Dockerfile (`lmsysorg/sglang:v0.5.12.post1-cu129`); resolved image digest and `pip freeze` attached to PR 2 before evidence runs |
| CPU / memory | 2 × Intel Xeon Platinum 8470Q, 208 logical CPUs / 1.0 TiB |
| GPU for capture evidence | 4 × NVIDIA RTX PRO 6000 Blackwell, 97,887 MiB each; driver 580.82.09 |
| Relax baseline | `8a54679e971087566bfda939a5f75649e07fb861` |

Offline core and fixtures require no model download. The two capture runs use the same pinned image, hardware, code and tensor/reduction settings; only the algorithm-specific GRPO/PPO configuration differs.

```bash
python -m pytest -v tests/utils/training/trajectory_replay/
python -m relax.tools.trajectory_replay verify --bundle tests/fixtures/trajectory_replay/valid-v1
python -m relax.tools.trajectory_replay run --stage all --bundle tests/fixtures/trajectory_replay/injected-mask-error
pre-commit run --all-files
```

## Reviewer question

Should PR 2 expose capture through a global CLI flag immediately, or land the Python capture API first and add CLI integration separately?

**Comment by SigureMo** [comment]
#223 入选，本 RFC close
